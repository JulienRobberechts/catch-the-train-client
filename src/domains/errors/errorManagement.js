import {
  ErrorMessages,
  errorInErrorManagementObject,
} from "./errorMessages.fr";
import ErrorLevels from "./errorLevels";
import { identifyError } from "./identifyError";

const handleError = (incomingError) => {
  try {
    const errorCode = identifyError(incomingError);
    const appError = getAppError(errorCode);
    LogErrorInternally(incomingError, appError);
    LogErrorForUser(appError);
    return toPublicError(appError);
  } catch (errorInErrorManagement) {
    console.log("Error in the error treatment", { errorInErrorManagement });
    return errorInErrorManagementObject;
  }
};

const toPublicError = (appError) => {
  const { code: errorCode, msg: errorMessage } = appError;

  const originalException =
    process.env.NODE_ENV !== "production"
      ? appError.originalException
      : undefined;

  return {
    errorType: "Error",
    errorCode,
    errorMessage,
    originalException,
  };
};

const getAppError = (errorCode) => {
  return ErrorMessages.find((e) => e.code === errorCode);
};

const LogErrorInternally = (errorInDev, errorObjectWithContext) => {
  if (
    process.env.NODE_ENV === "development" ||
    process.env.NODE_ENV === "test"
  ) {
    console.error({ errorInDev });
    console.error({ errorObjectWithContext });
  }

  // log via the server
  // todo ...
};

const LogErrorForUser = (errorObject) => {
  if (!errorObject) {
    console.error("Error management received an empty error");
    return;
  }

  const { code, level = ErrorLevels.MEDIUM, msg } = errorObject;
  const fullErrorMessage = `ERROR '${code}' (${level}) : '${msg}'`;

  // log on the client
  switch (errorObject.level) {
    case ErrorLevels.LOW:
      console.log(fullErrorMessage);
      break;
    case ErrorLevels.MEDIUM:
      console.warn(fullErrorMessage);
      break;
    case ErrorLevels.HIGH:
      console.error(fullErrorMessage);
      break;
    default:
      console.warn(fullErrorMessage);
      break;
  }
};

export { handleError, getAppError, LogErrorInternally, LogErrorForUser };
