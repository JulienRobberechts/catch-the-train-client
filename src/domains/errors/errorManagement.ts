import {
  ErrorMessages,
  errorInErrorManagementObject,
} from "./errorMessages.fr";
import ErrorLevels from "./errorLevels";
import { identifyError } from "./identifyError";
import { AppError, PublicError, AnyIncomingError } from "./types";

const handleError = (incomingError: AnyIncomingError): PublicError => {
  try {
    const errorCode = identifyError(incomingError);
    const appError = getAppError(errorCode);
    LogErrorInternally(incomingError, appError);
    LogErrorForUser(appError);
    return toPublicError(appError);
  } catch (errorInErrorManagement) {
    console.log("Error in the error treatment", { errorInErrorManagement });
    return toPublicError(errorInErrorManagementObject);
  }
};

const toPublicError = (appError: AppError): PublicError => {
  const { code: errorCode, msg: errorMessage } = appError;

  const originalException =
    process.env.NODE_ENV !== "production"
      ? appError.originalException
      : undefined;

  return {
    errorType: "Error",
    errorCode,
    errorMessage,
    originalException,
  };
};

const getAppError = (errorCode: number) : AppError => {
  const appError = ErrorMessages.find((e) => e.code === errorCode);
  return appError ?? errorInErrorManagementObject;
};

const LogErrorInternally = (incomingError: AnyIncomingError, errorObjectWithContext: any) => {
  if (
    process.env.NODE_ENV === "development" ||
    process.env.NODE_ENV === "test"
  ) {
    console.error({ incomingError });
    console.error({ errorObjectWithContext });
  }

  // log via the server
  // todo ...
};

const LogErrorForUser = (errorObject: AppError) => {
  if (!errorObject) {
    console.error("Error management received an empty error");
    return;
  }

  const { code, level = ErrorLevels.MEDIUM, msg } = errorObject;
  const fullErrorMessage = `ERROR '${code}' (${level}) : '${msg}'`;

  // log on the client
  switch (errorObject.level) {
    case ErrorLevels.LOW:
      console.log(fullErrorMessage);
      break;
    case ErrorLevels.MEDIUM:
      console.warn(fullErrorMessage);
      break;
    case ErrorLevels.HIGH:
      console.error(fullErrorMessage);
      break;
    default:
      console.warn(fullErrorMessage);
      break;
  }
};

export { handleError, getAppError, LogErrorInternally, LogErrorForUser };
