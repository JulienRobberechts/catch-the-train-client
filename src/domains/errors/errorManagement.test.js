import each from "jest-each";
import { handleError, getAppError } from "./errorManagement";
import errorCases from "./errorCaseTest";

describe("errorManagement", () => {
  // disable console logs for this test
  let originalError;
  let originalWarn;
  let originalLog;
  beforeAll(() => {
    originalError = console.error;
    console.error = jest.fn();
    originalWarn = console.warn;
    console.warn = jest.fn();
    originalLog = console.log;
    console.log = jest.fn();
  });
  afterAll(() => {
    console.error = originalError;
    console.warn = originalWarn;
    console.log = originalLog;
  });
  describe("handleError", () => {
    each`
      errorCaseId | errorCase                  
      ${1}        | ${errorCases.axiosError400}
      ${2}        | ${errorCases.axiosError500}
      ${3}        | ${errorCases.axiosError503}
      ${4}        | ${errorCases.simpleClientError700}
  `.test(
      "should handleError an error number $errorCaseId",
      ({ errorCaseId, errorCase }) => {
        const actualAppError = handleError(errorCase.incomingError);
        expect(actualAppError).toMatchSnapshot();
        expect(actualAppError.errorCode).toBeTruthy();
        expect(actualAppError.errorMessage).toBeTruthy();
        expect(actualAppError.errorCode).toEqual(errorCase.expectedErrorCode);
      }
    );

    test("should handleError an error REPRO", () => {
      const actualAppError = handleError(
        errorCases.simpleClientError700.incomingError
      );
      expect(actualAppError).toMatchSnapshot();
      expect(actualAppError.errorCode).toBeTruthy();
      expect(actualAppError.errorMessage).toBeTruthy();
      expect(actualAppError.errorCode).toEqual(
        errorCases.simpleClientError700.expectedErrorCode
      );
    });
  });

  describe("formatError", () => {
    each`
      errorCode   | expectedMessage
      ${503}      | ${"Le service est indisponible pour le moment"}
      ${700}      | ${"Erreur dans l'application (client web)"}
    `.test(
      "should format an error $errorCode",
      ({ errorCode, expectedMessage }) => {
        const actualObject = getAppError(errorCode);
        expect(actualObject.msg).toEqual(expectedMessage);
      }
    );
  });
});
