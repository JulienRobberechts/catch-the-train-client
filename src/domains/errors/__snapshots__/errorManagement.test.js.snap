// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errorManagement handleError should handleError an error REPRO 1`] = `
Object {
  "errorCode": 700,
  "errorMessage": "Erreur dans l'application (client web)",
  "errorType": "Error",
  "originalException": undefined,
}
`;

exports[`errorManagement handleError should handleError an error number 1 1`] = `
Object {
  "errorCode": 400,
  "errorMessage": "Erreur dans les donnees envoy√©s au serveur.",
  "errorType": "Error",
  "originalException": undefined,
}
`;

exports[`errorManagement handleError should handleError an error number 2 1`] = `
Object {
  "errorCode": 500,
  "errorMessage": "Une erreur inconnue s'est produite sur le serveur",
  "errorType": "Error",
  "originalException": undefined,
}
`;

exports[`errorManagement handleError should handleError an error number 3 1`] = `
Object {
  "errorCode": 503,
  "errorMessage": "L'application n'est pas disponible pour l'instant. Merci de re tenter plus tard.",
  "errorType": "Error",
  "originalException": undefined,
}
`;

exports[`errorManagement handleError should handleError an error number 4 1`] = `
Object {
  "errorCode": 700,
  "errorMessage": "Erreur dans l'application (client web)",
  "errorType": "Error",
  "originalException": undefined,
}
`;
